presto-hive/src/main/java/com/facebook/presto/hive/util/UgiUtils.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.util;
+
+import org.apache.hadoop.security.UserGroupInformation;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Utility class to handle creating and caching the UserGroupInformation object.
+ */
+public class UgiUtils
+{
+    private UgiUtils() {}
+
+    // Every instance of a UserGroupInformation object for a given user has a unique hashcode, due
+    // to the hashCode() impl. If we don't cache the UGI per-user here, there will be a memory leak
+    // in the PrestoFileSystemCache.
```

Comment:
Without this cache the specific leak would occur in ```PrestoFileSystemCache.map``` (https://github.com/prestodb/presto-hadoop-apache2/blob/master/src/main/java/org/apache/hadoop/fs/PrestoFileSystemCache.java#L62), since the Key would always be unique, even when the same user accesses the same filesystem in multiple queries.

Yes, the caching bounds the map size to O(number of users) by hashing UGIs consistently, but as implemented ```PrestoFileSystemCache.map``` has a "leak" in that it will cache all users keys for the duration of the JVM. Running our coordinator with 64G heap we ran into issues once we got to O(100k) map entries, to give a sense of the volume of users required to cause memory issues.